evm/src/cpu/kernel/asm/curve/bn254/curve_arithmetic/pairing.asm

```diff
@@ -0,0 +1,71 @@
+/// def bn254_pairing(pairs: List((Curve, TwistedCurve))) -> Fp12:
+///     
+///     for P, Q in pairs:
+///         assert(P.is_valid)
+///         assert(Q.is_valid)
+///     
+///     out = 1
+///     for P, Q in pairs:
+///         out *= miller_loop(P, Q)
+///
+///     return bn254_final_exponent(out)
+
+global bn254_pairing:
+    // stack: k, inp, out, retdest
+    DUP1
+
+bn254_input_check:
+    // stack:       j    , k, inp 
+    DUP1
+    ISZERO
+    // stack: end?, j    , k, inp
+    %jumpi(bn254_pairing_start)
+    // stack:       j    , k, inp
+    %sub_const(1)
+    // stack:       j=j-1, k, inp
+
+    %stack (j, k, inp) -> (j, inp, j, k, inp)
+    // stack:        j, inp, j, k, inp
+    %mul_const(6)
+    ADD
+    // stack:  inp_j=inp+6j, j, k, inp
+    DUP1
+    // stack:  inp_j, inp_j, j, k, inp
+    %load_fp254_2
+    // stack:    P_j, inp_j, j, k, inp
+    %bn_check
+    // stack: valid?, inp_j, j, k, inp
+    %assert_nonzero
```

Comment:
Here we shouldn't panic (which is what `assert_nonzero` does) but return a special value to indicate that the input was malformed. So you could replace this with `ISZERO %jumpi(bn_pairing_bad_input)` and 
```
bn_pairing_bad_input:
     %stack (...) -> (retdest, @U256_MAX)
     JUMP
```